src/noteburst/jupyterclient/jupyterlab.py

```diff
@@ -0,0 +1,708 @@
+        ----------
+        final : `bool`
+            The last attempt, so log some additional information if the lab
+            still isn't gone.
+        """
+        user_url = self.url_for(f"hub/api/users/{self.user.username}")
+        referer_url = self.url_for("hub/home")
+        headers = {"Referer": referer_url}
+        r = await self.http_client.get(user_url, headers=headers)
+        if r.status_code != 200:
+            raise JupyterError.from_response(self.user.username, r)
+        data = r.json()
+        result = data["servers"] == {}
+        if final and not result:
+            msg = f'Server data still shows running lab: {data["servers"]}'
+            self.logger.warning(msg)
+        return result
+
+    @contextlib.asynccontextmanager
+    async def open_lab_session(
+        self, notebook_name: Optional[str] = None, *, kernel_name: str = "LSST"
+    ) -> AsyncGenerator[JupyterLabSession, None]:
+        """Open a JupyterLab session.
+
+        Send and receive messages from JupyterLab using the ``websocket``
+        property on `JupyterLabSession`.
+        """
+        session_url = self.url_for(f"user/{self.user.username}/api/sessions")
+        session_type = "notebook" if notebook_name else "console"
+        body = {
+            "kernel": {"name": kernel_name},
+            "name": notebook_name or "(no notebook)",
+            "path": notebook_name if notebook_name else uuid4().hex,
+            "type": session_type,
+        }
+        r = await self.http_client.post(session_url, json=body)
+        if r.status_code != 201:
+            raise JupyterError.from_response(self.user.username, r)
+        session_resource = r.json()
+
+        kernel_id = session_resource["kernel"]["id"]
+        http_channels_uri = self.url_for(
+            f"user/{self.user.username}/api/kernels/{kernel_id}/channels"
+        )
+        wss_channels_uri = self.url_for_websocket(
+            f"user/{self.user.username}/api/kernels/{kernel_id}/channels"
+        )
+
+        # Generate a mock request and copy its headers / cookies over to the
+        # websocket connection.
+        mock_request = self.http_client.build_request("GET", http_channels_uri)
+        copied_headers = ["x-xsrftoken", "authorization", "cookie"]
+        websocket_headers = {
+            header: mock_request.headers[header] for header in copied_headers
+        }
+
+        session_id: Optional[str] = None  # will be set if a session is opened
+        self.logger.debug("Trying to create websocket connection")
+        try:
+            # An alternative to the async context it to use connect in an
+            # infinite async generator that can automatically reconnect
+            # if the connection is dropped. This could be good for very
+            # long lived clients
+            # https://websockets.readthedocs.io/en/stable/reference/client.html#using-a-connection
+            async with websockets.connect(  # type: ignore
```

Comment:
Out of curiosity, what was mypy unhappy about here?

Reply:
I get

```
src/noteburst/jupyterclient/jupyterlab.py:642: error: Module has no attribute "connect"  [attr-defined]
```

I think it's because websockets has an unusual way of importing `connect` into that namespace.